From 79b53e5420519d52c637bfa2c0523bf9cf30748a Mon Sep 17 00:00:00 2001
From: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Date: Tue, 1 Feb 2022 10:28:33 +0200
Subject: [PATCH 5/6] pmic: add board-specific pmic operations

Add board-specific pmic operations and create a separate header for
common s32g pmic operations.

Issue: ALB-8443
Upstream-Status: Pending 

Signed-off-by: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32/s32_psci.c                       |  1 +
 plat/nxp/s32/s32g/include/s32g_bl_common.h    |  6 ------
 plat/nxp/s32/s32g/include/s32g_vr5510.h       | 18 +++++++++++++++++
 plat/nxp/s32/s32g/s32g2/s32g2.mk              |  1 -
 .../s32/s32g/s32g2/s32g274ardb2/platform.mk   |  2 ++
 .../s32g274ardb2_vr5510.c}                    |  9 ++++++++-
 .../s32/s32g/s32g2/s32g2xxaevb/platform.mk    |  2 ++
 .../s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c    | 20 +++++++++++++++++++
 plat/nxp/s32/s32g/s32g3/s32g3_vr5510.c        |  9 ++++++++-
 plat/nxp/s32/s32g/s32g_bl2_el3.c              |  1 +
 plat/nxp/s32/s32g/s32g_resume.c               |  1 +
 plat/nxp/s32/s32g/s32g_vr5510.c               | 15 +++++++++++---
 12 files changed, 73 insertions(+), 12 deletions(-)
 create mode 100644 plat/nxp/s32/s32g/include/s32g_vr5510.h
 rename plat/nxp/s32/s32g/s32g2/{s32g2_vr5510.c => s32g274ardb2/s32g274ardb2_vr5510.c} (87%)
 create mode 100644 plat/nxp/s32/s32g/s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c

diff --git a/plat/nxp/s32/s32_psci.c b/plat/nxp/s32/s32_psci.c
index ca2ee418e..24c2c44de 100644
--- a/plat/nxp/s32/s32_psci.c
+++ b/plat/nxp/s32/s32_psci.c
@@ -18,6 +18,7 @@
 #include "s32g_clocks.h"
 #include "s32g_mc_me.h"
 #include "s32g_resume.h"
+#include "s32g_vr5510.h"
 #else
 #include "s32_bl_common.h"
 #include "s32_mc_me.h"
diff --git a/plat/nxp/s32/s32g/include/s32g_bl_common.h b/plat/nxp/s32/s32g/include/s32g_bl_common.h
index a3d7dfc84..b396cecf2 100644
--- a/plat/nxp/s32/s32g/include/s32g_bl_common.h
+++ b/plat/nxp/s32/s32g/include/s32g_bl_common.h
@@ -6,16 +6,10 @@
 #ifndef S32G_BL_COMMON_H
 #define S32G_BL_COMMON_H
 
-#include <pmic/vr5510.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include "s32_bl_common.h"
 
-int pmic_prepare_for_suspend(void);
-void pmic_system_off(void);
-int pmic_disable_wdg(vr5510_t fsu);
-int pmic_setup(void);
-
 void s32g_reinit_i2c(void);
 
 bool s32gen1_is_wkp_short_boot(void);
diff --git a/plat/nxp/s32/s32g/include/s32g_vr5510.h b/plat/nxp/s32/s32g/include/s32g_vr5510.h
new file mode 100644
index 000000000..8170fe4bb
--- /dev/null
+++ b/plat/nxp/s32/s32g/include/s32g_vr5510.h
@@ -0,0 +1,18 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef S32G_VR5510_H
+#define S32G_VR5510_H
+
+#include <pmic/vr5510.h>
+#include <stdint.h>
+
+int pmic_prepare_for_suspend(void);
+void pmic_system_off(void);
+int pmic_disable_wdg(vr5510_t fsu);
+int pmic_setup(void);
+uint16_t pmic_stby_pwr_rails(void);
+
+#endif /* S32G_VR5510_H */
diff --git a/plat/nxp/s32/s32g/s32g2/s32g2.mk b/plat/nxp/s32/s32g/s32g2/s32g2.mk
index f1ee8e0ff..701a81f42 100644
--- a/plat/nxp/s32/s32g/s32g2/s32g2.mk
+++ b/plat/nxp/s32/s32g/s32g2/s32g2.mk
@@ -20,7 +20,6 @@ PLAT_BL_COMMON_SOURCES	+= drivers/nxp/s32/clk/s32g274a_clk.c \
 			   plat/nxp/s32/s32gen1_mc_me.c \
 			   plat/nxp/s32/s32gen1_mc_rgm.c \
 			   plat/nxp/s32/s32gen1_sramc.c \
-			   plat/nxp/s32/s32g/s32g2/s32g2_vr5510.c \
 			   lib/cpus/aarch64/s32.S \
 			   lib/cpus/aarch64/cortex_a53.S \
 
diff --git a/plat/nxp/s32/s32g/s32g2/s32g274ardb2/platform.mk b/plat/nxp/s32/s32g/s32g2/s32g274ardb2/platform.mk
index addf145f0..093543c84 100644
--- a/plat/nxp/s32/s32g/s32g2/s32g274ardb2/platform.mk
+++ b/plat/nxp/s32/s32g/s32g2/s32g274ardb2/platform.mk
@@ -8,4 +8,6 @@ include plat/nxp/s32/s32g/s32g2/s32g2.mk
 
 PLAT_INCLUDES		+= -Iplat/nxp/s32/s32g/s32g2/s32g274ardb2/include \
 
+PLAT_BL_COMMON_SOURCES	+=	plat/nxp/s32/s32g/s32g2/s32g274ardb2/s32g274ardb2_vr5510.c \
+
 DTB_FILE_NAME		?= fsl-s32g274a-rdb2.dtb
\ No newline at end of file
diff --git a/plat/nxp/s32/s32g/s32g2/s32g2_vr5510.c b/plat/nxp/s32/s32g/s32g2/s32g274ardb2/s32g274ardb2_vr5510.c
similarity index 87%
rename from plat/nxp/s32/s32g/s32g2/s32g2_vr5510.c
rename to plat/nxp/s32/s32g/s32g2/s32g274ardb2/s32g274ardb2_vr5510.c
index f36db639d..5a45bcc2b 100644
--- a/plat/nxp/s32/s32g/s32g2/s32g2_vr5510.c
+++ b/plat/nxp/s32/s32g/s32g2/s32g274ardb2/s32g274ardb2_vr5510.c
@@ -1,11 +1,18 @@
 /*
- * Copyright 2021 NXP
+ * Copyright 2021-2022 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 #include <s32g_bl_common.h>
+#include <s32g_vr5510.h>
 #include <drivers/nxp/s32/ocotp.h>
 
+uint16_t pmic_stby_pwr_rails(void)
+{
+	return (VR5510_CTRL3_VPREV_STBY | VR5510_CTRL3_HVLDO_STBY
+	      | VR5510_CTRL3_BUCK3_STBY | VR5510_CTRL3_LDO2_STBY);
+}
+
 static int is_svs_needed(bool *status)
 {
 	uint32_t val;
diff --git a/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/platform.mk b/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/platform.mk
index d8cf828dd..1257e1d59 100644
--- a/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/platform.mk
+++ b/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/platform.mk
@@ -8,4 +8,6 @@ include plat/nxp/s32/s32g/s32g2/s32g2.mk
 
 PLAT_INCLUDES		+= -Iplat/nxp/s32/s32g/s32g2/s32g2xxaevb/include \
 
+PLAT_BL_COMMON_SOURCES	+=	plat/nxp/s32/s32g/s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c \
+
 DTB_FILE_NAME		?= fsl-s32g2xxa-evb.dtb
\ No newline at end of file
diff --git a/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c b/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c
new file mode 100644
index 000000000..1e37d5c62
--- /dev/null
+++ b/plat/nxp/s32/s32g/s32g2/s32g2xxaevb/s32g2xxaevb_vr5510.c
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <s32g_vr5510.h>
+
+int pmic_prepare_for_suspend(void)
+{
+	return 0;
+}
+
+void pmic_system_off(void)
+{}
+
+int pmic_disable_wdg(vr5510_t fsu)
+{
+	return 0;
+}
diff --git a/plat/nxp/s32/s32g/s32g3/s32g3_vr5510.c b/plat/nxp/s32/s32g/s32g3/s32g3_vr5510.c
index f14596e59..59e9a8dc1 100644
--- a/plat/nxp/s32/s32g/s32g3/s32g3_vr5510.c
+++ b/plat/nxp/s32/s32g/s32g3/s32g3_vr5510.c
@@ -1,9 +1,16 @@
 /*
- * Copyright 2021 NXP
+ * Copyright 2021-2022 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 #include <s32g_bl_common.h>
+#include <s32g_vr5510.h>
+
+uint16_t pmic_stby_pwr_rails(void)
+{
+	return (VR5510_CTRL3_VPREV_STBY | VR5510_CTRL3_HVLDO_STBY
+	      | VR5510_CTRL3_BUCK3_STBY | VR5510_CTRL3_BUCK1_STBY);
+}
 
 int pmic_setup(void)
 {
diff --git a/plat/nxp/s32/s32g/s32g_bl2_el3.c b/plat/nxp/s32/s32g/s32g_bl2_el3.c
index 129cdcf4e..dc74b2ce9 100644
--- a/plat/nxp/s32/s32g/s32g_bl2_el3.c
+++ b/plat/nxp/s32/s32g/s32g_bl2_el3.c
@@ -13,6 +13,7 @@
 #include "bl31_ssram.h"
 #include "s32_bl2_el3.h"
 #include "s32g_bl_common.h"
+#include "s32g_vr5510.h"
 #include <plat/nxp/s32g/bl31_ssram/ssram_mailbox.h>
 #include "s32_sramc.h"
 #if S32G_EMU == 1
diff --git a/plat/nxp/s32/s32g/s32g_resume.c b/plat/nxp/s32/s32g/s32g_resume.c
index 583083c7b..6110d36c1 100644
--- a/plat/nxp/s32/s32g/s32g_resume.c
+++ b/plat/nxp/s32/s32g/s32g_resume.c
@@ -7,6 +7,7 @@
 #include "s32_linflexuart.h"
 #include "s32g_lowlevel.h"
 #include "s32g_resume.h"
+#include "s32g_vr5510.h"
 #include "s32gen1-wkpu.h"
 #include <bl31/bl31.h>		/* for bl31_warm_entrypoint() */
 #include <lib/el3_runtime/context_mgmt.h>
diff --git a/plat/nxp/s32/s32g/s32g_vr5510.c b/plat/nxp/s32/s32g/s32g_vr5510.c
index a25c62841..81eb999aa 100644
--- a/plat/nxp/s32/s32g/s32g_vr5510.c
+++ b/plat/nxp/s32/s32g/s32g_vr5510.c
@@ -1,9 +1,15 @@
 /*
- * Copyright 2019-2021 NXP
+ * Copyright 2019-2022 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 #include <s32g_bl_common.h>
+#include <s32g_vr5510.h>
+
+#pragma weak pmic_prepare_for_suspend
+#pragma weak pmic_system_off
+#pragma weak pmic_disable_wdg
+#pragma weak pmic_setup
 
 static int watchdog_refresh(vr5510_t fsu)
 {
@@ -98,8 +104,7 @@ int pmic_prepare_for_suspend(void)
 	if (ret)
 		return ret;
 
-	reg = VR5510_CTRL3_VPREV_STBY | VR5510_CTRL3_HVLDO_STBY
-		| VR5510_CTRL3_BUCK3_STBY |  VR5510_CTRL3_LDO2_STBY;
+	reg = pmic_stby_pwr_rails();
 	ret = vr5510_write(mu, VR5510_M_REG_CTRL3, regp, sizeof(reg));
 	if (ret)
 		return ret;
@@ -254,3 +259,7 @@ int pmic_disable_wdg(vr5510_t fsu)
 	return watchdog_refresh(fsu);
 }
 
+int pmic_setup(void)
+{
+	return 0;
+}
-- 
2.17.1

